[Bug]: Windows 平台正常退出依然会生成崩溃日志

### 在提问之前...

- [X] 我理解 Issue 是用于反馈和解决问题的，而非吐槽评论区，将尽可能提供更多信息帮助问题解决
- [X] 我填写了简短且清晰明确的标题，以便开发者在翻阅 issue 列表时能快速确定大致问题。而不是“一个建议”、“卡住了”等
- [X] 我已查看master branch或最新测试版的更新内容，并未提及该 bug 已被修复的情况
- [X] 已有issue中并没有看见其他人与我反馈相同的问题

### 问题描述

正常方式退出程序（点击托盘菜单中的“退出程序”）依然会生成崩溃报告

### 如何复现

1. 启动海豹
2. 等待几分钟
3. 右键右下角退出程序
4. 复现

### 你期望发生的

不生成崩溃报告

### 实际上发生的

生成了崩溃报告

### 日志文件

```
2024-04-09T17:39:14.901-0400    info    dice/platform_adapter_kook.go:352       KOOK 连接成功，账号<SzzRain的测试机器人>(KOOK:2210765268)       {"dice": "default"}
2024-04-09T17:39:14.903-0400    info    dice/config.go:2388     保存数据        {"dice": "default"}
2024-04-09T20:23:32.359-0400    info    sealdice-core/main.go:44        程序即将退出，进行清理……
2024-04-09T20:23:32.388-0400    info    sealdice-core/main.go:44        程序即将退出，进行清理……
panic: close of closed channel [recovered]
        panic: close of closed channel

goroutine 1 [running, locked to thread]:
sealdice-core/dice.CrashLog()
        C:/Users/MSI-NB/GolandProjects/sealdice-core/dice/dice.go:756 +0x174
panic({0x7ff667ddd000, 0x7ff6686bb430})
        C:/Users/MSI-NB/go/go1.20.2/src/runtime/panic.go:890 +0x263
github.com/blevesearch/bleve/v2/index/scorch.(*Scorch).Close(0xc0020f7b00)
        C:/Users/MSI-NB/go/pkg/mod/github.com/blevesearch/bleve/v2@v2.3.10/index/scorch/scorch.go:332 +0xa5
github.com/blevesearch/bleve/v2.(*indexImpl).Close(0xc001fc4070)
        C:/Users/MSI-NB/go/pkg/mod/github.com/blevesearch/bleve/v2@v2.3.10/index_impl.go:852 +0xa2
sealdice-core/dice.(*HelpManager).Close(0xc0018f8180)
        C:/Users/MSI-NB/GolandProjects/sealdice-core/dice/dice_help.go:161 +0x36
main.cleanUpCreate.func1()
        C:/Users/MSI-NB/GolandProjects/sealdice-core/main.go:112 +0x310
main.main()
        C:/Users/MSI-NB/GolandProjects/sealdice-core/main.go:429 +0x1517

Process finished with the exit code 2
```

### 截图

_No response_

### 海豹核心版本

https://github.com/sealdice/sealdice-core/commit/85a79999351f4de8e58b6a12b15cf1e5143f5674

### 操作系统

Windows

### 帐号类型

Kook（按理说无关）

### 使用协议

_No response_

### 附加内容

_No response_

--- a/dice/dice_manager.go
+++ b/dice/dice_manager.go
@@ -3,6 +3,7 @@ package dice
 import (
 	"fmt"
 	"os"
+	"sync/atomic"
 	"time"
 
 	"github.com/dop251/goja_nodejs/require"
@@ -75,7 +76,8 @@ type DiceManager struct { //nolint:revive
 	JsRegistry           *require.Registry
 	UpdateSealdiceByFile func(packName string, log *zap.SugaredLogger) bool // 使用指定压缩包升级海豹，如果出错返回false，如果成功进程会自动结束
 
-	ContainerMode bool // 容器模式：禁用内置适配器，不允许使用内置Lagrange和旧的内置Gocq
+	ContainerMode bool          // 容器模式：禁用内置适配器，不允许使用内置Lagrange和旧的内置Gocq
+	CleanupFlag   atomic.Uint32 // 1 为正在清理，0为普通状态
 }
 
 type DiceConfigs struct { //nolint:revive

--- a/go.mod
+++ b/go.mod
@@ -53,6 +53,7 @@ require (
 	github.com/glebarez/go-sqlite v1.22.0
 	github.com/go-creed/sat v1.0.3
 	github.com/golang-module/carbon v1.7.3
+	github.com/gonutz/w32/v2 v2.11.1
 	github.com/gorilla/websocket v1.5.1
 	github.com/grokify/html-strip-tags-go v0.0.1
 	github.com/jmoiron/sqlx v1.4.0

--- a/main.go
+++ b/main.go
@@ -41,7 +41,7 @@ data/logs
 extensions/
 */
 
-func cleanUpCreate(diceManager *dice.DiceManager) func() {
+func cleanupCreate(diceManager *dice.DiceManager) func() {
 	return func() {
 		logger.Info("程序即将退出，进行清理……")
 		err := recover()
@@ -52,6 +52,11 @@ func cleanUpCreate(diceManager *dice.DiceManager) func() {
 			if runtime.GOOS == "windows" {
 				exec.Command("pause") // windows专属
 			}
+		}
+
+		if !diceManager.CleanupFlag.CompareAndSwap(0, 1) {
+			// 尝试更新cleanup标记，如果已经为1则退出
+			return
 		}
 
 		for _, i := range diceManager.Dice {
@@ -393,7 +398,7 @@ func main() {
 
 	go dice.TryGetBackendURL()
 
-	cleanUp := cleanUpCreate(diceManager)
+	cleanUp := cleanupCreate(diceManager)
 	defer dice.CrashLog()
 	defer cleanUp()
 

--- a/tray_darwin.go
+++ b/tray_darwin.go
@@ -86,7 +86,7 @@ func onReady() {
 			_ = exec.Command(`open`, filepath.Dir(os.Args[0])).Start()
 		case <-mQuit.ClickedCh:
 			systrayQuited = true
-			cleanUpCreate(theDm)()
+			cleanupCreate(theDm)()
 			systray.Quit()
 			time.Sleep(3 * time.Second)
 			os.Exit(0)

--- a/tray_windows.go
+++ b/tray_windows.go
@@ -143,7 +143,7 @@ func onReady() {
 		case <-mQuit.ClickedCh:
 			systray.Quit()
 			systrayQuited = true
-			cleanUpCreate(theDM)()
+			cleanupCreate(theDM)()
 			time.Sleep(3 * time.Second)
 			os.Exit(0)
 		case <-mAutoBoot.ClickedCh:

--- a/update.go
+++ b/update.go
@@ -117,7 +117,7 @@ func doReboot(dm *dice.DiceManager) {
 	}
 	platform := runtime.GOOS
 	if platform == "windows" {
-		cleanUpCreate(dm)()
+		cleanupCreate(dm)()
 
 		// name, _ := filepath.Abs(binary)
 		// err = exec.Command(`cmd`, `/C`, "start", name, "--delay=15").Start()
@@ -128,7 +128,7 @@ func doReboot(dm *dice.DiceManager) {
 		}
 	} else {
 		// 手动cleanup
-		cleanUpCreate(dm)()
+		cleanupCreate(dm)()
 		// os.Args[1:]...
 		execErr := syscall.Exec(binary, []string{os.Args[0], "--delay=25"}, os.Environ())
 		if execErr != nil {

--- a/update_updater.go
+++ b/update_updater.go
@@ -189,7 +189,7 @@ func UpdateByFile(dm *dice.DiceManager, log *zap.SugaredLogger, packName string,
 				}
 
 				time.Sleep(5 * time.Second)
-				cleanUpCreate(dm)()
+				cleanupCreate(dm)()
 				os.Exit(0)
 			}
 			if syncMode {

